x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlTasks.java

```diff
@@ -223,6 +225,30 @@ public static DataFrameAnalyticsState getDataFrameAnalyticsState(@Nullable Persi
         return state;
     }
 
+    public static TrainedModelDeploymentState getTrainedModelDeploymentState(PersistentTasksCustomMetadata.PersistentTask<?> task) {
+        if (task == null) {
+            return TrainedModelDeploymentState.STOPPED;
+        }
+        TrainedModelDeploymentTaskState taskState = (TrainedModelDeploymentTaskState) task.getState();
+        if (taskState == null) {
+            return TrainedModelDeploymentState.STARTING;
+        }
+
+        TrainedModelDeploymentState state = taskState.getState();
+        if (taskState.isStatusStale(task)) {
+            if (state == TrainedModelDeploymentState.STOPPING) {
+                // previous executor node failed while the job was stopping - it won't
+                // be restarted on another node, so consider it STOPPED for reassignment purposes
+                return TrainedModelDeploymentState.STOPPED;
+            }
+            if (state != TrainedModelDeploymentState.FAILED) {
+                // we are relocating at the moment
+                return TrainedModelDeploymentState.STARTING;
+            }
```

Comment:
It's probably worth adding a TODO here to revisit for the new-style model allocation because, unlike persistent task, that won't have the same concept of "relocating".